[TOPIC-GPIO] updates for deprecated API

pabigot opened the pull request:
This PR captures the deprecation portion of the topic-gpio merge checklist in #21789.  Note that some commits will be squashed between this and other tasks before being merged.

pabigot committed:
gpio: deprecate legacy gpio pin read/write functions

These have been replaced by get/set functions in both raw and
active-sensitive variants.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
gpio: replace gpio pin write/read with set/get

This API will be deprecated in favor of the new API that clearly
specifies whether it works on logical or physical levels.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
SQUASH: tests: gpio_basic_api: remove test for deprecated functions

To be merged with the commit that deprecates; currently separated to
support manual verification that the deprecation substitutions work.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
gpio: replace GPIO_DIR flags

GPIO_DIR_OUT is now GPIO_OUTPUT, similarly for GPIO_DIR_IN.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nashif reviewed (dismissed):
Can you please avoid "tree-wide:" and "treewide" in commit messages, this is obviously related to gpio, so why not prefix it with "gpio:"?

pabigot commented:
> Can you please avoid "tree-wide:" and "treewide" in commit messages, this is obviously related to gpio, so why not prefix it with "gpio:"?

Because it's not modifying gpio-specific files.  In an automated fashion it's modifying board setup functions and bit-banging I2C, and anything else anywhere in the tree that use the functions.  But I'll change it anyway.

pabigot committed:
boards: replace GPIO_PUD_PULL flags in devicetree source

GPIO_PUD_PULL_UP is now spelled GPIO_PULL_UP.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_cc13xx_cc26xx: replace GPIO_PUD_PULL macros

GPIO_PUD_PULL_foo is now spelled GPIO_PULL_foo.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
gpio: replace GPIO_INT_LEVEL flag

This will be deprecated, use the generalized mode flag or absence of
GPIO_INT_EDGE.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
gpio: mark deprecated flag macros

Mark the outdated flag macros as deprecated so they can be fixed.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: bluetooth: hci: replace deprecated gpio_pin callback API

gpio_pin_en/disable_callback are to be replaced by
gpio_pin_interrupt_configure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: led: ht16k33: reorder pin interrupt configuration

Replace the enable callback API with the enable interrupt API, and do
it after the callback is installed.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: sensor: lis3mdl: replace deprecated gpio_pin callback API

gpio_pin_disable_callback is to be replaced by
gpio_pin_interrupt_configure.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
west.yml: update mcuboot to use non-deprecated GPIO flags

Upcoming deprecation will produce build failures as long as the
deprecated flags are used.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

avisconti reviewed (approved):
ooops... sorry!

pabigot committed:
doc: release-notes-2.2: update for GPIO deprecations

Add a section describing what's been deprecated and how to replace it.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
gpio: deprecate gpio pin callback enable and disable API

These have been replaced by the appropriate call to
gpio_pin_interrupt_configure().  While the disable function could be
implemented using the new functionality, the enable function cannot
because the interrupt mode is not available.  Consequently we cannot
replace these with equivalent functionality using the legacy API.

Clean up the internal implementation by removing the inaccessible
port-based enable/disable feature, leaving the pin-based capability in
place.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: gpio_basic_api: add tests of deprecated functionality

Since the callback enable and disable functions are deprecated, but
cannot be re-implemented in terms of other API, add back an old-style
interrupt configuration test that uses the deprecated flags and
functions in the way existing code would do.  The test module
internally disables the deprecation warnings.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nashif commented:
> Because it's not modifying gpio-specific files.

it is about context, not necessarily files. Every change in this series is related to gpio. thanks for changing.

carlescufi reviewed (approved):
I see enabling the deprecation actually made quite a few missing changes visible!
This is great work, I think the release notes update is very well written too.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -1341,9 +1352,10 @@ static inline int gpio_remove_callback(struct device *port,
  * interrupt configuration flags such as ``GPIO_INT_EDGE_TO_ACTIVE``.
  */
+/* Deprecated in 2.2 release */
 static inline int gpio_pin_enable_callback(struct device *port, u32_t pin)
```
`__deprecated`

mnkp commented on include/drivers/gpio.h:
```diff
@@ -1355,9 +1367,10 @@ static inline int gpio_pin_enable_callback(struct device *port, u32_t pin)
  * ``GPIO_INT_DISABLE``.
  */
+/* Deprecated in 2.2 release */
 static inline int gpio_pin_disable_callback(struct device *port, u32_t pin)
```
`__deprecated`

mnkp reviewed (changes requested):
@pabigot that was a lot of work. Thanks! PR looks good overall but I have some comments, a few blocking.

The deprecated `gpio_pin_write` function should be implemented in terms of `gpio_pin_set`. Shortly, I'll add an explanation why.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -1217,19 +1230,15 @@ static inline int gpio_pin_toggle(struct device *port, unsigned int pin)
-		 "Unsupported pin");
-
-	return gpio_write(port, GPIO_ACCESS_BY_PIN, pin, value);
+	return gpio_pin_set_raw(port, pin, value != 0);
```
Several of the existing GPIO drivers support GPIO_POL_INV flag. An examples are: gpio_cc13xx_cc26xx.c, gpio_esp32.c, gpio_mmio32.c, gpio_pcal9535a.c, gpio_sifive.c, gpio_sx1509b.c, gpio_nrfx.c. The effect of passing the GPIO_POL_INV flag is inversion of the pin level. If the GPIO_POL_INV flag was passed to the `gpio_pin_configure` function subsequent calling of `gpio_pin_write(,,1)` will result in pin being driven low. That behavior can only be correctly implemented with help of `gpio_pin_set`. Any out of tree application that is currently using GPIO_POL_INV flag and one of the affected drivers will break if `gpio_pin_set_raw` is used. We need to preserve backwards compatibility.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -1242,19 +1251,21 @@ static inline int gpio_pin_write(struct device *port, u32_t pin,
 {
-	const struct gpio_driver_config *const cfg =
-		(const struct gpio_driver_config *)port->config->config_info;
+	int rv = gpio_pin_get_raw(port, pin);
```
Should be implemented using `gpio_pin_get`

mnkp commented:
There is one more issue which we probably should mention in the release notes. Several drivers the rely on the functionality provided by the GPIO API, here is the list:
- drivers/bluetooth/hci/spi.c
- drivers/display/display_ili9340.c
- drivers/display/ssd1306.c
- drivers/eeprom/eeprom_at2x.c
- drivers/ieee802154/ieee802154_mcr20a.c
- drivers/ieee802154/ieee802154_rf2xx_iface.c
- drivers/lora/sx1276.c
- drivers/sensor/ccs811/ccs811.c
- drivers/wifi/eswifi/eswifi_core.c
were converted to use pin logical levels. That means that any out of the tree users will need to update their DTS to correctly configure active level of the pin or the drivers will not work. I need to double check if all of the above drivers are affected or only some. Most likely only a few, those that have reset signal which is typically active low.

Should we mention this in the release notes?

pabigot commented on doc/releases/release-notes-2.2.rst:
```diff
@@ -31,6 +31,36 @@ Deprecated in this release
+    * GPIO_INT_LEVEL becomes not GPIO_INT_EDGE
+    * GPIO_INT_ACTIVE_LOW becomes GPIO_ACTIVE_LOW or GPIO_INT_LOW_0
+    * GPIO_INT_ACTIVE_HIGH becomes GPIO_ACTIVE_HIGH or GPIO_INT_HIGH_1
+    * GPIO_INT_DOUBLE_EDGE becomes GPIO_INT_EDGE_BOTH
```
I'm going to leave this as is; you or others can tweak it when this is in master.  I suggest keeping track of this comment as reference.

In fact, I think you should be the one to write the release notes, since it's your design and vision we've implemented.

pabigot commented on boards/arm/96b_wistrio/pinmux.c:
```diff
@@ -62,17 +62,17 @@ static int pinmux_stm32_init(struct device *port)
-	gpio_pin_configure(gpioa, 4, GPIO_DIR_OUT);
-	gpio_pin_write(gpioa, 4, 1);
+	gpio_pin_configure(gpioa, 4, GPIO_OUTPUT);
+	gpio_pin_set_raw(gpioa, 4, 1);
```
We could, but this was an automated global replace.   There are multiple places this cleanup could be done (this commit and others), but doing that is manual, so I'll leave it for normal maintenance.

pabigot committed:
SQUASH gpio: replace gpio pin write/read with set/get (logic)

Revises previous commit to use the logic level API instead of the
physical level API.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
SQUASH gpio: deprecate legacy gpio pin read/write functions (logic)

Revises previous commit to use the logic level API instead of the
physical level API.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
PR updated to have commits that convert write/read to set/get instead of the raw variants.  This resulted in some previously separate conversions being squashed for convenience.

Missed deprecation flags added.

Other than a mis-named function no release note content has been changed.  What's there was intended to capture the changes I made; I've updated #21789 to make release notes the responsibility of @mknp as the new API implements his vision.

mnkp reviewed (approved):
LGTM, apart from one extra non-blocking comment regarding an issue I've overlooked before.

The commit that re-implements `gpio_pin_write` functionality in terms of `gpio_pin_set` should ideally remove `gpio_write` syscall and `write` function from the api structure as well as all references thereof in the driver's code. The same should be done with `read`. As it is now, we have some dead code in the tree. Of course, we can clean it up later but logically it belongs here. I'll leave the decision to @pabigot.
